python/lsst/multiprofit/modeller.py

```diff
@@ -0,0 +1,840 @@
+        results.result = result_opt
+        if params_free_sorted_missing:
+            params_best = []
+            for param in params_free_sorted_all:
+                if param in params_free_sorted_missing:
+                    params_best.append(param.value)
+                    param.fixed = False
+                else:
+                    params_best.append(result_opt.x[offsets_params[param] - 1])
+            results.params_best = tuple(params_best)
+        else:
+            results.params_best = tuple(
+                result_opt.x[offsets_params[param] - 1] for param in params_free_sorted
+            )
+        results.n_eval_func = result_opt.nfev
+        results.n_eval_jac = result_opt.njev if result_opt.njev else 0
+
+        return results
+
+    # TODO: change to staticmethod if requiring py3.10+
+    @classmethod
+    def fit_model_linear(
+        cls,
+        model: Model,
+        idx_obs: int = None,
+        ratio_min: float = 0,
+        validate: bool = False,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        n_data = len(model.data)
+        n_sources = len(model.sources)
+        if n_sources != 1:
+            raise ValueError("fit_model_linear does not yet support models with >1 sources")
+        if idx_obs is not None:
+            if not ((idx_obs >= 0) and (idx_obs < n_data)):
+                raise ValueError(f"{idx_obs=} not >=0 and < {len(model.data)=}")
+            indices = range(idx_obs, idx_obs + 1)
+        else:
+            indices = range(n_data)
+
+        if validate:
+            model.setup_evaluators(evaluatormode=g2f.EvaluatorMode.loglike)
+            loglike_init = model.evaluate()
+        else:
+            loglike_init = None
+        values_init = {}
+        values_new = {}
+
+        for idx_obs in indices:
+            obs = model.data[idx_obs]
+            gaussians_linear = LinearGaussians.make(model.sources[0], channel=obs.channel)
+            result = cls.fit_gaussians_linear(gaussians_linear, obs, psf_model=model.psfmodels[idx_obs])
+            values = list(result.values())[0]
+
+            for (_, parameter), ratio in zip(gaussians_linear.gaussians_free, values):
+                values_init[parameter] = float(parameter.value)
+                if not (ratio >= ratio_min):
+                    ratio = ratio_min
+                value_new = max(ratio * parameter.value, parameter.limits.min)
+                values_new[parameter] = value_new
+
+        for parameter, value in values_new.items():
+            # TODO: maybe just np.clip instead
+            if not (value > parameter.limits.min):
+                value = parameter.limits.min + (
+                    1e-5
+                    if (parameter.limits.max == np.Inf)
+                    else (0.02 * (parameter.limits.max - parameter.limits.min))
+                )
+            elif not (value < parameter.limits.max):
+                value = parameter.limits.min + 0.98 * (parameter.limits.max - parameter.limits.min)
```

Comment:
I refactored this and added some docs but the summary is that this is meant to avoid linear fits setting fluxes to zero which, depending on the parameter's transform, can make it difficult or impossible for subsequent nonlinear fits to move above zero.